internal/route/api/v1/repo/contents.go

```diff
@@ -0,0 +1,177 @@
+		return
+	}
+
+	// TODO: figure out the best way to do this
+	// :base-url/:username/:project/raw/:refs/:path
+	templateDownloadURL := "%s/%s/%s/raw/%s"
+	// :base-url/repos/:username/:project/contents/:path
+	templateSelfLink := "%s/repos/%s/%s/contents/%s"
+	// :baseurl/repos/:username/:project/git/trees/:sha
+	templateGitURLLink := "%s/repos/%s/%s/trees/%s"
+	// :baseurl/repos/:username/:project/tree/:sha
+	templateHTMLLLink := "%s/repos/%s/%s/tree/%s"
+
+	gitURL := fmt.Sprintf(templateGitURLLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), treeEntry.ID.String())
+	htmlURL := fmt.Sprintf(templateHTMLLLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), treeEntry.ID.String())
+	selfURL := fmt.Sprintf(templateSelfLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), c.Repo.TreePath)
+
+	contents := &repoContents{
+		Size:        treeEntry.Size(),
+		Name:        treeEntry.Name(),
+		Path:        c.Repo.TreePath,
+		Sha:         treeEntry.ID.String(),
+		URL:         selfURL,
+		GitURL:      gitURL,
+		HTMLURL:     htmlURL,
+		DownloadURL: fmt.Sprintf(templateDownloadURL, c.BaseURL, c.Params(":username"), c.Params(":reponame"), c.Repo.TreePath),
+		Links: Links{
+			Git:  gitURL,
+			Self: selfURL,
+			HTML: htmlURL,
+		},
+	}
+
+	// A tree entry can only be one of the following types:
+	// 1. Tree (Directory)
+	// 2. SubModule
+	// 3. SymLink
+	// 4. Blob
+	if treeEntry.IsSubModule() {
+		contents.Type = "submodule"
+		parsedURL, err := url.Parse(c.BaseURL)
+		if err != nil {
+			c.ServerError("ErrorURLParse", err)
+			return
+		}
+		host := parsedURL.Host
+		submoduleURL := fmt.Sprintf("git://%s/%s/%s.git", host, c.Params(":name"), c.Params(":reponame"))
+		contents.SubmoduleGitURL = submoduleURL
+		c.JSONSuccess(contents)
+		return
+
+	} else if treeEntry.IsLink() {
+		contents.Type = "symlink"
+		contents.Target = c.Repo.TreePath
+		c.JSONSuccess(contents)
+		return
+
+	} else if treeEntry.Type == "blob" {
+		blob, err := c.Repo.Commit.GetBlobByPath(c.Repo.TreePath)
+		if err != nil {
+			c.ServerError("ErrorGetBlobByPath", err)
+			return
+		}
+		buf := make([]byte, 1024)
+		b, err := blob.Data()
+		if err != nil {
+			c.ServerError("BlobDataError", err)
+			return
+		}
+		n, err := b.Read(buf)
+
+		if err != nil {
+			c.ServerError("ContentReadError", err)
+			return
+		}
+		if n >= 0 {
+			buf = buf[:n]
+		}
+		contents.Content = base64.StdEncoding.EncodeToString(buf)
+		contents.Type = "file"
+		c.JSONSuccess(contents)
+		return
+	}
+
+	// treeEntry is a directory
+	dirTree, err := c.Repo.GitRepo.GetTree(treeEntry.ID.String())
+	if err != nil {
+		c.NotFoundOrServerError("GetGitDirTree", git.IsErrNotExist, err)
+		return
+	}
+
+	entries, err := dirTree.ListEntries()
+	if err != nil {
+		c.NotFoundOrServerError("ListDirTreeEntries", git.IsErrNotExist, err)
+		return
+	}
+
+	var results = make([]*repoContents, 0, len(entries))
+	if len(entries) == 0 {
+		c.JSONSuccess(&repoGitTree{})
```

Comment:
Does GitHub API response an empty Git tree for empty directory? What do we get from a GitHub API in this case?

Reply:
Good question, let me check that.

Reply:
So @unknwon , I tried creating a empty directory, but looks like you cannot commit an empty directory to git. Also tried committing a file along and later deleting that file to make the directory empty - also not possible.

 But with an empty repo, when I tested I got the following response:
```bash
curl -H "Authorization: token REDACTED" https://api.github.com/repos/bharatnc/test-repo/contents | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   132  100   132    0     0    261      0 --:--:-- --:--:-- --:--:--   261
{
  "message": "This repository is empty.",
  "documentation_url": "https://developer.github.com/v3/repos/contents/#get-contents"
}
```
I think we're almost guaranteed not to get an empty dir, but if we do, we should just respond with a server error I think. Does that sound right to you?

Reply:
Thanks for confirming this, yep, SGTM. Or simpler, just an empty array in JSON `[]`, because it is empty.

Reply:
Sure just returning an empty array `c.JSONSuccess([]string{})`.